packages/tasit-action/src/Contract.js

```diff
@@ -0,0 +1,542 @@
+    } else if (eventName === "error") {
+      this._addErrorListener(listener);
+    } else if (eventName === "confirmation") {
+      this.#addConfirmationListener(listener, once);
+    }
+  };
+
+  #addConfirmationListener = (listener, once) => {
+    const eventName = "confirmation";
+
+    const baseEthersListener = async blockNumber => {
+      try {
+        if (!this.#tx) this.#tx = await this.#txPromise;
+
+        const receipt = await this.#provider.getTransactionReceipt(
+          this.#tx.hash
+        );
+
+        const blockReorgOccurred =
+          (receipt === null && this.#txConfirmations > 0) ||
+          (receipt !== null && receipt.confirmations <= this.#txConfirmations);
+
+        if (blockReorgOccurred) {
+          this._emitErrorEvent(
+            new Error(
+              `Your action's position in the chain has changed in a surprising way.`
+            ),
+            eventName
+          );
+        }
+
+        if (receipt === null) {
+          this.#txConfirmations = 0;
+          return;
+        }
+
+        this._clearEventTimerIfExists(eventName);
+
+        this.#lastConfirmationTime = Date.now();
+
+        const timer = setTimeout(() => {
+          const currentTime = Date.now();
+          const timedOut =
+            currentTime - this.#lastConfirmationTime >= this.getEventsTimeout();
+
+          if (timedOut) {
+            this._emitErrorEvent(
+              new Error(`Event ${eventName} reached timeout.`),
+              eventName
+            );
+          }
+        }, this.getEventsTimeout());
+
+        this._setEventTimer(eventName, timer);
+
+        const { confirmations } = receipt;
+
+        this.#txConfirmations = confirmations;
+
+        const message = {
+          data: {
+            confirmations,
+          },
+        };
+
+        await listener(message);
+      } catch (error) {
+        this._emitErrorEvent(
+          new Error(`Listener function with error: ${error.message}`),
+          eventName
+        );
+      }
+    };
+
+    let ethersListener;
+
+    if (once) {
+      ethersListener = async blockNumber => {
+        await baseEthersListener(blockNumber);
+        this.off(eventName);
+      };
+    } else {
+      ethersListener = baseEthersListener;
+    }
+
+    this._addEventListener(eventName, ethersListener);
+  };
+
+  // Tech debt
+  // This method avoids duplicated nonce generation when several transactions happen in rapid succession
+  // See: https://github.com/ethereumbook/ethereumbook/blob/04f66ae45cd9405cce04a088556144be11979699/06transactions.asciidoc#keeping-track-of-nonces
+  // How should we keep track of nonces?
+  waitForNonceToUpdate = async () => {
+    const tx = await this.#txPromise;
+    await this.#provider.waitForTransaction(tx.hash);
+  };
+
+  // For testing purposes
+  refreshProvider = () => {
+    this.#provider = ProviderFactory.getProvider();
+  };
+}
+
+class ContractSubscription extends Subscription {
```

Comment:
I'm in doubt about this change because with an object of this class the user can register to N events of this contract. The name `ContractEventSubscription' may suggests the this class represents one subscription on a  contract event. What do you think?

Reply:
Grammatically speaking, the fact that it's singular here doesn't necessarily imply that it's only for a single event - but I agree that it would be far less confusing if it didn't have that possible interpretation in the first place.

The reason I think ContractSubscription isn't enough is because sending an NFT is basically changing state on a single contract, so if I were "subscribed" to that contract, I sure would hope that I would hear about it. So I think we need to show that this subclass is only related to the events that come from the contract, not transactions involving the contract.

Maybe `ContractEventsSubscription`. The two s's in a row don't look that nice, but it's the most concise wording I can think of.

Reply:
```suggestion
class ContractEventsSubscription extends Subscription {
```

Reply:
Thank you for explanation. I've renamed on last commit since the code is outdated.

Reply:
Great. Resolving this.